Fix compilation, add stubbed out version of missing source files.

OscarL opened the pull request:
All copies of TraX sources I could find (including the ones on this repo), are incomplete. They miss at least the following files (along some missing docs, like `settings.html`):

- `BarberPole.{h,cpp}`.
- `Globals.h`
- `Joins.h`
- `ResultsHandler.{h,cpp}`
- `Settings.h` (.cpp too?)
- `SettomgsWindow.{h,cpp}`

(BeOS binaries for TraX/FastTraX also have mentions of a `KeyFilter.cpp` file)


For BarberPole, I've used [BarberPole1.0.zip](https://pulkomandy.tk/~beosarchive/index.pl?search=barberpole) from PulkoMandy's BeOS Archive.

For the rest, I just did my best at trying to supply stubbed-out versions (see last commit).

On my 3rd attempt (first two long ago), finally managed to make TraX compile on Haiku beta4, 32 bits, with GCC2.

~**NOTE**: It crashes on initialization of my stubbed `ResultsHandler` class/object, thou!~

~Not unexpected, considering my lack of C/C++ skills. Help welcomed!~

Seems to work now.

----

I tried to keep commits more or less "atomic", just in case someone wants to cherry-pick instead of merge them all.

Not sure I'll be able to do much more with this one (lack of skills and focus). Maybe someone else can "fill in the blanks" / fix what remains broken?

We'll see :-)

OscarL committed:
Add stubbed-out versions of missing source files.

Should at least compile/link cleanly now (on 32bits, GCC2).

Both TraX and FastTraX seem functional (except the Settings window).

OscarL committed:
Add Barberpole.{h.cpp} from BarberPole1.0.zip

Got the file from https://pulkomandy.tk/~beosarchive/.

Changes to that:

* Renamed BarberPole.cc -> BarberPole.cpp..
* Removed include for "Colors.h", no idea where that comes from
  and doesn't seems actually needed.
* Added default value for the flag constructor parameter.
  Not sure if the value is correct, but prevents compile errors
  for now.

OscarL committed:
Fix permissions for .sh scripts.

OscarL committed:
Fix includes (missing or wrong).

OscarL committed:
Fix BeOS/Haiku defines.

OscarL committed:
dfa.c: Fix most warnings in this file.

OscarL committed:
Define these so the correct code paths are used.

Made sure config.h gets included first.

OscarL committed:
No sysmacros.h in Haiku, AFAICT.

OscarL committed:
B_ENTRY_NOT_FOUND was deprecated on hrev55532.

OscarL committed:
Make sure `Common/lib` exists.

Otherwise linking of libTraX.so fails.

Surely there's a better place to do this mkdir?

OscarL committed:
Add missing forward declaration.

diversys commented:
On x86_64 build fails with the following:
```
~/TraX> make
gcc -o FindLib/lib/libfind.so  FindLib/obj.r1/contains.o  FindLib/obj.r1/pred.o  FindLib/obj.r1/express.o  FindLib/obj.r1/util.o  FindLib/obj.r1/findlib.o  FindLib/obj.r1/error.o  FindLib/obj.r1/tree.o  FindLib/obj.r1/xgetcwd.o  FindLib/obj.r1/modechange.o  FindLib/obj.r1/xmalloc.o  FindLib/obj.r1/xstrdup.o  FindLib/obj.r1/fstype.o  FindLib/obj.r1/savedir.o  FindLib/obj.r1/filemode.o  FindLib/obj.r1/listfile.o  FindLib/obj.r1/idcache.o FRegex/lib/fregex.a  -shared -Xlinker -soname=libfind.so
/boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/13.2.0/../../../../x86_64-unknown-haiku/bin/ld: FindLib/obj.r1/pred.o:(.bss+0x0): multiple definition of `fast_regex_subsys_error'; FindLib/obj.r1/contains.o:(.bss+0x0): first defined here
/boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/13.2.0/../../../../x86_64-unknown-haiku/bin/ld: FindLib/obj.r1/express.o:(.bss+0x0): multiple definition of `fast_regex_subsys_error'; FindLib/obj.r1/contains.o:(.bss+0x0): first defined here
/boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/13.2.0/../../../../x86_64-unknown-haiku/bin/ld: FindLib/obj.r1/util.o:(.bss+0x0): multiple definition of `fast_regex_subsys_error'; FindLib/obj.r1/contains.o:(.bss+0x0): first defined here
/boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/13.2.0/../../../../x86_64-unknown-haiku/bin/ld: FindLib/obj.r1/findlib.o:(.bss+0x10): multiple definition of `xstat'; FindLib/obj.r1/express.o:(.data.rel+0x0): first defined here
/boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/13.2.0/../../../../x86_64-unknown-haiku/bin/ld: FindLib/obj.r1/findlib.o:(.bss+0x78): multiple definition of `fast_regex_subsys_error'; FindLib/obj.r1/contains.o:(.bss+0x0): first defined here
/boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/13.2.0/../../../../x86_64-unknown-haiku/bin/ld: FindLib/obj.r1/tree.o:(.bss+0x0): multiple definition of `fast_regex_subsys_error'; FindLib/obj.r1/contains.o:(.bss+0x0): first defined here
/boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/13.2.0/../../../../x86_64-unknown-haiku/bin/ld: FindLib/obj.r1/fstype.o:(.bss+0x0): multiple definition of `fast_regex_subsys_error'; FindLib/obj.r1/contains.o:(.bss+0x0): first defined here
/boot/system/develop/tools/bin/../lib/gcc/x86_64-unknown-haiku/13.2.0/../../../../x86_64-unknown-haiku/bin/ld: FRegex/lib/fregex.a(fast_regex.o):(.bss+0x0): multiple definition of `fast_regex_subsys_error'; FindLib/obj.r1/contains.o:(.bss+0x0): first defined here
collect2: error: ld returned 1 exit status
FindLib/etc/makefile.rules:7: recipe for target 'FindLib/lib/libfind.so' failed
make: *** [FindLib/lib/libfind.so] Error 1
~/TraX> 


```

OscarL commented:
Hi @diversys, thanks for checking. 64 bits, beta4 or nightly (just to rule things out).

I've only tested in 32 bits (without even using setarch). Will look into it anyway.

Thanks again.


Edit: for now... getting `gcc: error: unrecognized command-line option '-nostart'` here on beta4 64 bits instead.


After changing that `-nostart` to `-shared`, yeah. I get the same error.

diversys commented:
Always latest nightly. 
> Edit: for now... getting gcc: error: unrecognized command-line option '-nostart' here on beta4 64 bits instead.

Yeah, forgot to mention that I did the same thing.

OscarL commented:
Commenting out line 37 on `FRegex/lib/fast_regex.h` seems to avoid that error, but then I get:

```multiple definition of `xstat'; FindLib/obj.r1/express.o:(.data.rel+0x0): first defined here```

And a lack of wits prevents further progress on my part :-D

diversys commented:
Managed to compile it with these changes:

```
diff --git a/Common/Support/Support.h b/Common/Support/Support.h
index f6dc83f..933e5db 100644
--- a/Common/Support/Support.h
+++ b/Common/Support/Support.h
@@ -60,13 +60,13 @@ namespace ObjectUtils
 {
 	// Duplicates an object from another pointer
 	template <class T>
-	inline T* Dup( T* )
+	inline T* Dup( T* t )
 	{
 		return new T( *t ) ;
 	}  
 
 	template <class T>
-	inline void Swap( T& , T& )
+	inline void Swap( T& a, T& b )
 	{
 		T c = a ;
 		a = b ;
diff --git a/Common/Support/WindowList.cpp b/Common/Support/WindowList.cpp
index da84531..2cc25b9 100644
--- a/Common/Support/WindowList.cpp
+++ b/Common/Support/WindowList.cpp
@@ -93,7 +93,7 @@ void WindowList :: CloseAll( void )
 {
 	Lock() ;
 	void * v ;
-	while( (v = fList.RemoveItem( 0L )) != NULL )
+	while( (v = fList.RemoveItem( (int32)0L )) != NULL )
 	{
 		reinterpret_cast<BWindow *>( v )->Quit() ;
 	}
diff --git a/Common/UI/ResultsWindow.cpp b/Common/UI/ResultsWindow.cpp
index b5f530b..872cc45 100644
--- a/Common/UI/ResultsWindow.cpp
+++ b/Common/UI/ResultsWindow.cpp
@@ -89,7 +89,7 @@ ResultsWindow :: ResultsWindow( void )
 ResultsWindow :: ~ResultsWindow( void )
 {
 	BListItem * bli ;
-	while(  (bli = fpList->RemoveItem(0L)) != NULL )
+	while ((bli = fpList->RemoveItem(static_cast<int32>(0L))) != NULL)
 		delete bli ;
 }
 
@@ -132,7 +132,7 @@ bool ResultsWindow :: QuitRequested( void )
 		}
 	}
 	
-	return inherited::QuitRequested() ;
+	return QuitRequested() ;
 }
 
 void ResultsWindow :: MessageReceived( BMessage * msg )
@@ -153,7 +153,7 @@ void ResultsWindow :: MessageReceived( BMessage * msg )
 
 		default:
 		{
-			inherited::MessageReceived( msg ) ;
+			MessageReceived( msg ) ;
 			break ;
 		}
 	}
diff --git a/FRegex/lib/fast_regex.h b/FRegex/lib/fast_regex.h
index 8445d91..2b7bce8 100644
--- a/FRegex/lib/fast_regex.h
+++ b/FRegex/lib/fast_regex.h
@@ -34,7 +34,6 @@ struct fast_regex
 
 typedef struct fast_regex fast_regex_t ;
  
-int (*fast_regex_subsys_error)( const char * sys, const char * msg ) ;
 
 int  allocate_fast_regex ( fast_regex_t * ) ;
 int  compile_fast_regex  ( fast_regex_t *, const char *, size_t ) ;
diff --git a/FastTrax/UIMain/FastTraxWindow.cpp b/FastTrax/UIMain/FastTraxWindow.cpp
index 9a41000..bb19473 100644
--- a/FastTrax/UIMain/FastTraxWindow.cpp
+++ b/FastTrax/UIMain/FastTraxWindow.cpp
@@ -419,7 +419,7 @@ bool FastTraxWindow::QuitRequested( void )
 			msgr.SendMessage( &msg , &reply ) ;
 		}
 	}
-	return inherited :: QuitRequested( ) ;
+	return  QuitRequested( ) ;
 }
 
 // -1 Kill
@@ -625,7 +625,7 @@ void FastTraxWindow :: MessageReceived( BMessage * msg )
 		}		
 		default:
 		{
-			inherited::MessageReceived( msg ) ;
+			MessageReceived( msg ) ;
 		}
 	}
 }
diff --git a/FindLib/etc/makefile b/FindLib/etc/makefile
index d6c207b..ad17994 100644
--- a/FindLib/etc/makefile
+++ b/FindLib/etc/makefile
@@ -13,7 +13,7 @@ CFLAGS			=	$(OPTIMIZER) -Wall -Wno-multichar -Wno-ctor-dtor-privacy
 
 LD				=	gcc
 LDFLAGS			=	
-SHAREDLIB_FLAGS	=	-nostart -Xlinker -soname=$(notdir $@)
+SHAREDLIB_FLAGS	=	-shared -Xlinker -soname=$(notdir $@)
 
 
 
diff --git a/FindLib/findlib/express.c b/FindLib/findlib/express.c
index 7a2df71..6e13523 100644
--- a/FindLib/findlib/express.c
+++ b/FindLib/findlib/express.c
@@ -41,7 +41,6 @@ static int insert_fprintf_main ( FILE * fp, boolean ( *func ) () , char * fmt )
 static int insert_days ( enum comparison_type c_type , unsigned long num_days, PFB pred ) ;
 static int insert_time ( enum comparison_type c_type , time_t t , PFB pred ) ;
 
-int ( *xstat ) () = lstat ;
 
 #define ISDIGIT(c) (isascii (c) && isdigit (c))
 #define ISUPPER(c) (isascii (c) && isupper (c))
diff --git a/etc/makefile.main b/etc/makefile.main
index dc44b00..f490d6d 100644
--- a/etc/makefile.main
+++ b/etc/makefile.main
@@ -23,7 +23,7 @@ CFLAGS			 =	$(OPTIMIZER) -Wall -Wno-multichar -Wno-ctor-dtor-privacy -D__USE_GNU
 
 LD				:=	gcc
 LDFLAGS			:=	
-SHAREDLIB_FLAGS	 = -nostart -Xlinker -soname=$(notdir $@)
+SHAREDLIB_FLAGS	 = -shared -Xlinker -soname=$(notdir $@)
 APP_FLAGS        = -Xlinker -soname=_APP_
 
 XRES		    := xres
```

Unfortunately, it crashes with 

```
	thread 3921: team 3920 debug task 
	thread 3920: TraX (main)
		state: Call (You need a valid BApplication object before interacting with the app_server)

		Frame		IP			Function Name
		-----------------------------------------------
		00000000	0xa21057	_kern_debugger + 0x7 
			Disassembly:
				_kern_debugger:
				0x0000000000a21050:   48c7c0e9000000  mov $0xe9, %rax
				0x0000000000a21057:             0f05  syscall  <--

		0x7f0001082ea0	0x83048f	BWindow::_InitData(BRect, char const*, window_look, window_feel, unsigned int, unsigned int, int) + 0x74f 
		0x7f0001082f20	0x8305c1	BWindow::BWindow(BRect, char const*, window_type, unsigned int, unsigned int) + 0xa1 
		0x7f0001082f70	0x9b23f6	ResultsWindow::ResultsWindow() + 0x36 
		0x7f0001082f90	0x9b46cc	ResultsHandler::ResultsHandler() + 0xc 
		0x7f0001082fb0	0x60d068	_GLOBAL(int16, , bool, , , ) + 0x18 
		0x7f0001083060	0x107d83	init_dependencies(image_t*, bool) [clone .constprop.0] + 0x1c3 
		0x7f00010830f0	0x1086b7	load_program + 0x1d7 
		0x7f0001083120	0x10fa6e	runtime_loader + 0xee

OscarL commented:
Thanks a lot Diver!

Yeah... that's the same crash I get on 32 bits.

The culprit is one of the (stubbed out) files/classes I've added :-/

Seems I got it correct enough to compile/link (with GCC2 at least), but more than wrong enough to make it crash :-D

I have really no idea what I was doing with the `gSettings` and `gResults` globals objects that were missing. Was all pure uneducated guesses, so it would have been a miracle it started properly.

Might be that ResultsHandler should not inherit from ResultsWindow, but just use an instance of that (by composition) instead. That should help avoid trying to create a BWindow before we even got be_app up and running (will try a bit more on the weekend).

Thanks again!

(still wishing for someone to show up with a correct version of the missing classes :-D)

OscarL commented:
(Couldn't resist trying a bit more...)

Alright! Just managed to make it compile AND make `FastTraX` start now (still on 32 bits GCC2). It even shows results! TraX still crashes thou.

Will rework my commits a bit, and push them soon.

Edit: TraX seems to work now :D

OscarL committed:
Fixes for GCC13.

Both FastTraX and Trax build, start, and seem to function ok on
beta4 32 bits, AND on 64 bits (GCC13).

(still missing Settings/SettingsWindow functionality, of course).

Co-Authored-by: Sergei Reznikov <diver@gelios.net>

OscarL committed:
Fix most warning about integer formats.

korli commented on Common/Support/Scripter.cpp:
```diff
@@ -138,7 +138,7 @@ int32 Scripter :: FindMatch( BMessage * m, int32 i,
 
 	if( fOptions & PrintDebug )
-		fprintf( stderr, "Match for %s is %ld (%p)\n", prop, idx, this ) ;
+		fprintf( stderr, "Match for %s is %" B_PRIi32 " (%p)\n", prop, idx, this ) ;
```
```suggestion
		fprintf( stderr, "Match for %s is %" B_PRId32 " (%p)\n", prop, idx, this ) ;
```

korli commented on Common/UI/EditBox.cpp:
```diff
@@ -126,28 +126,28 @@ void EditBox :: SetFloat( float f , const char * fmt )
 {
 	char str[64] ;
-	sprintf( str, "%ld", i ) ;
+	sprintf( str, "%" B_PRIi32, i ) ;
```
```suggestion
	sprintf( str, "%" B_PRId32, i ) ;
```

korli commented on FRegex/source/dfa.c:
```diff
@@ -625,11 +625,12 @@ lex()
 				{
 					setbit(c, ccl);
-					if (case_fold)
+					if (case_fold) {
```
```suggestion
					if (case_fold)
					{
```

korli commented on FRegex/source/dfa.c:
```diff
@@ -1849,11 +1850,12 @@ int *backref;
 			if (d->success[s] & sbit[*p])
 			{
-				if (backref)
+				if (backref) {
```
```suggestion
				if (backref)
				{
```

korli commented on FRegex/source/fast_regex.c:
```diff
@@ -2,6 +2,7 @@
 
 #include <stdio.h>
 #include <stdio.h>
+#include <stdlib.h>
```
```suggestion
#include <stdlib.h>
```

OscarL replied:
Heh, I didn't even noticed that `stdio.h` was there 3 times originally :-)

Will fix.

korli commented on FindLib/findlib/express.c:
```diff
@@ -24,6 +24,8 @@
 #include <sys/stat.h>
 #include <grp.h>
 #include <pwd.h>
+#include <posix/ctype.h>
+#include <posix/regex.h>
```
any reasons for posix prefix?

OscarL replied:
None other than my lack of expertise.

Thanks for the review, BTW! Will address your suggestions after some sleep. Thanks again!

korli commented on TraxSource/UIMain/OptionsPanel.cpp:
```diff
@@ -245,7 +245,7 @@ void OptionsPanel :: StartFind( void )
 	{
 		int32 i = fpMaxDepthEdit->GetInt() ;
-		printf("MaxDepth: %ld\n", i ) ;
+		printf("MaxDepth: %" B_PRIi32 "\n", i ) ;
```
```suggestion
		printf("MaxDepth: %" B_PRId32 "\n", i ) ;
```

korli commented on etc/makefile.main:
```diff
@@ -19,11 +19,12 @@ FASTTRAX_DIR := FastTrax
 
 CC		     	:=	gcc
-CFLAGS			 =	$(OPTIMIZER) -Wall -Wno-multichar -Wno-ctor-dtor-privacy
+CFLAGS			 =	$(OPTIMIZER) -Wall -Wno-multichar -Wno-ctor-dtor-privacy -D__USE_GNU -DSTDC_HEADERS -D_POSIX_SOURCE
```
```suggestion
CFLAGS			 =	$(OPTIMIZER) -Wall -Wno-multichar -Wno-ctor-dtor-privacy -D_DEFAULT_SOURCE -DSTDC_HEADERS -D_POSIX_SOURCE
```

Usually setting the right -std option is better. see https://gcc.gnu.org/onlinedocs/gcc/C-Dialect-Options.html
For instance -std=gnu11

OscarL replied:
I dropped `-DSTDC_HEADERS` (defined already on `Findlib/findlib/config.h`). Works, but generates more warnings down the line (on code under `FRegex/`).

I replaced `-D_POSIX_SOURCE` with a define in the mentioned `config.h` file.

Using -D_DEFAULT_SOURCE instead of -D__USE_GNU gives me (on GCC13):

```bash
FindLib/findlib/express.c: In function 'insert_regex_main':
FindLib/findlib/express.c:670:13: error: 'struct re_pattern_buffer' has no member named 'allocated'; did you mean '__allocated'?
  670 |         re->allocated = 100 ;
      |             ^~~~~~~~~
      |             __allocated
FindLib/findlib/express.c:671:13: error: 'struct re_pattern_buffer' has no member named 'buffer'; did you mean '__buffer'?
  671 |         re->buffer = ( unsigned char * ) xmalloc ( re->allocated ) ;
      |             ^~~~~~
      |             __buffer
FindLib/findlib/express.c:671:56: error: 'struct re_pattern_buffer' has no member named 'allocated'; did you mean '__allocated'?
  671 |         re->buffer = ( unsigned char * ) xmalloc ( re->allocated ) ;
      |                                                        ^~~~~~~~~
      |                                                        __allocated
FindLib/findlib/express.c:672:13: error: 'struct re_pattern_buffer' has no member named 'fastmap'; did you mean '__fastmap'?
  672 |         re->fastmap = NULL ;
      |             ^~~~~~~
      |             __fastmap
FindLib/findlib/express.c:678:21: error: 'struct re_pattern_buffer' has no member named 'translate'; did you mean '__translate'?
  678 |                 re->translate = xmalloc ( 256 ) ;
      |                     ^~~~~~~~~
      |                     __translate
FindLib/findlib/express.c:681:29: error: 'struct re_pattern_buffer' has no member named 'translate'; did you mean '__translate'?
  681 |                         re->translate[ i ] = ISUPPER ( i ) ? tolower ( i ) : i ;
      |                             ^~~~~~~~~
      |                             __translate
FindLib/findlib/express.c:684:21: error: 'struct re_pattern_buffer' has no member named 'translate'; did you mean '__translate'?
  684 |                 re->translate = NULL ;
      |                     ^~~~~~~~~
      |                     __translate

```

Same when trying to use `-D_GNU_SOURCE` instead. Tried with `-std=gnu90`, `-std=gnu9x`, `-std=gnu++98`... same result.

For `__USE_GNU`:

I can't just define it in the mentioned `Findlib/findlib/config.h`, as several files under `FRegex/source/` also needs it.

I guess I could use a `#{define,undef} __USE_GNU` block when using `#include <regex.h>`, but that seems far more intrusive. (the whole project needs a serious clean up of its includes and header guards).

OscarL replied:
I'll stick with just `-D__USE_GNU -DSTDC_HEADERS` for this PR. Might not be ideal, but the alternatives *I* can come up with are just more convoluted/intrusive, and require far more work on the `#includes` than I want to do at this point.

(I rather try helping to add the missing settings functionality).

Can we leave it for future improvement? :-D

korli replied:
Maybe include features.h then it should define __USE_GNU as needed.

OscarL replied:
I'm feeling dumber than usual... `#include <features.h>` doesn't seems to help, and none of the three `features.h` files I find with `query -a features.h | grep system` defines (or mentions) `__USE_GNU`.

Grepping the entire `develop/headers` for `__USE_GNU` shows no defines for it.

I see it being defined in Haiku's [`src/system/libroot/posix/glibc/include/features.h`](https://cgit.haiku-os.org/haiku/tree/src/system/libroot/posix/glibc/include/features.h#n234) (guarded by `#ifdef _GNU_SOURCE`), but that header doesn't seems present under `/system/develop` on my beta4 install.

I *did* tried  it like this:

```c++
#define _GNU_SOURCE
#include <features.h>
```

and couldn't make it work (while still need to use `-DSTDC_HEADERS` and `#define _POSIX_SOURCE` in `Findlib/findlib/config.h`).

Sorry if I'm missing something obvious... Goes to show how little I know, and makes me wish I had a different hobby :-/

korli replied:
Right, _GNU_SOURCE should imply __USE_GNU, but that missing in features.h

pulkomandy replied:
I would take a different approach. `__USE_GNU` is glibc specific and I didn't expect it would be used in public headers. So headers/posix/regex.h should be adapted to use our variant of features.h, which use internal defines that are not compatible with glibc.

It seems this header is the only one in headers/posix that was imported directly from glibc? (or maybe there are more, but nothing that uses `__USE_GNU`). Usually our public headers are written by ourselves, and glibc headers, if we use any, are kept private (in sources/system/libroot) and not available for applications.

OscarL replied:
Sounds good to me, but I guess that means waiting for beta5 for this PR (at least for TraX to reach HaikuPorts)... or...

Maybe the `-D__USE_GNU` on CFLAGS is good enough here in the meantime? (maybe while adding a `ToDo` comment on the `makefile.main`?

pulkomandy replied:
yes, whichever fix we do in Haiku in the end, it will probably be only for the next release anyway. Until then, I think defining `__USE_GNU` here is fine.

diversys commented:
I think Setting button should be hidden as right now clicking it opens an empty window which can't be closed and the app had to be killed.

OscarL committed:
Disable the "Settings" button, until we reimplement the needed bits.

Avoids having to kill the app due to the current non-closable
SettingsWindow.

diversys commented:
@pulkomandy @korli is this ready to be merged?

OscarL commented:
I've added a comment over https://dev.haiku-os.org/ticket/11818#comment:7. Maybe that ticket should be re-open to keep track of the Haiku side of things?

OscarL committed:
Allow compilation on both beta4 and nightlies with a fixed features.h

OscarL commented:
After hrev57485 (and the follow up fixes on hrev57497) fixed `feature.h`, we can now use it for nightlies, but just in case, I've added detection of the proper hrev, so we might try and build this in current beta4 buildmasters too.
